### DISCUSSION TITLE
k3s pre-releases version mismatch with upstream version

### DISCUSSION THREAD

**Discussion by discordianfish** [k3s-io/k3s#7562] [Q&A] [answered]
**Environmental Info:**
K3s Version: v1.27.1-rc1+k3s1

**Describe the bug:**
k3s v1.27.1-rc1+k3s1 sets the server version to 1.27.1-rc1 but that upstream k8s version doesn't exist. This breaks assumptions of e.g sonobuoy that there is a conformance image 1.27.1-rc1 which there is not.
 
**Steps To Reproduce:**
- Install k3s v1.27.1-rc1+k3s1
- Run `sonobuoy run` and check e2e job failing with `Failed to pull image "registry.k8s.io/conformance:v1.27.1-rc1":`

**Expected behavior:**
Dunno, maybe it can be argued that sonobuoy can't use the server version to find the right conformance image. Or maybe k3s needs their own conformance image that is release lockstep with k3s release.
But I feel the easist and correct fix would be marking k3s release candiates by suffixing -rc at the end, so in this case the version should have been named `v1.27.1+k3s1-rc1`

**Comment by brandond** [comment, marked as answer]
Yes, this was raised as a concern internally when we switched versioning schemes and started releasing K3s using upstream Kubernetes versioning, and were trying to decide how to tag our release candidates - should we put the `rc` slug in the build metadata, or in the pre-release metadata?

While the current approach does make our release candidate versions indistinguishable from upstream's release candidate versions, it also makes tools that do semver comparisons able to properly identify our release candidates as having pre-release metadata, and not consider them full releases.

If we did it the other way (`v1.27.1+k3s1-rc1` for example), the pre-release metadata would be hidden in the build metadata, and not properly identifiable by most tools.

For this reason, we are not planning on changing our tagging conventions.

For what you're trying to do here, sonobuoy has a flag (`--kubernetes-version`) to override the detected version; I suggest you use that when working with our RCs. That is far and away the "easiest and correct fix", compared to asking us to change our long-standing processes and break semver compatibility.

**Comment by discordianfish** [comment]
Sure, I mean it's not a problem for me. It just took me a while to debug and felt like it's worth raising. I would just assume that the version reported in the k8s api would have some conformance constraints that might be violated. But very well possible that it's not and tools like sonobuoy can't expect that the reported server version matches an upstream release.
